File: auto_permissions/providers/factory.py
"""Factory function for initializing and configuring provider hierarchies."""
from typing import Any, Dict

from .base import BaseProvider
from .openai_compat import OpenAICompatibleProvider
from .ollama import OllamaProvider
from .gemini import GeminiProvider
from .tiered import TieredProvider

def get_provider(config: Dict[str, Any]) -> BaseProvider:
    """Instantiate and wire up providers based on config."""
    provider_name = config.get("provider", "llamacpp").lower()
    endpoint = config.get("endpoint", "http://127.0.0.1:9931/v1/chat/completions")
    model = config.get("model", "auto")
    timeout = float(config.get("timeout_seconds", 3.5))
    temperature = float(config.get("temperature", 0.0))
    num_ctx = int(config.get("num_ctx", 4096))

    # Construct primary provider
    if provider_name in ("llamacpp", "openai") or "v1/chat/completions" in endpoint:
        primary = OpenAICompatibleProvider(
            endpoint=endpoint,
            model=model,
            temperature=temperature,
            timeout=timeout,
        )
    elif provider_name == "gemini":
        api_key = config.get("api_key") or config.get("gemini_api_key")
        return GeminiProvider(
            api_key=api_key,
            model=model if model not in ("auto", "default") else "gemini-flash-lite-latest",
            temperature=temperature,
            timeout=timeout,
        )
    else:
        primary = OllamaProvider(
            endpoint=endpoint,
            model=model if model not in ("auto", "default") else "qwen3.5:9b",
            num_ctx=num_ctx,
            temperature=temperature,
            timeout=timeout,
        )

    # Check for cloud failover
    if config.get("fallback_to_cloud", True):
        cloud_model = config.get("cloud_model", "gemini-flash-lite-latest")
        cloud_timeout = float(config.get("cloud_timeout_seconds", 4.5))
        cloud_api_key = config.get("gemini_api_key") or config.get("api_key")
        secondary = GeminiProvider(
            api_key=cloud_api_key,
            model=cloud_model,
            temperature=temperature,
            timeout=cloud_timeout,
        )
        total_deadline = float(config.get("total_deadline_seconds", 11.0))
        return TieredProvider(primary, secondary, total_deadline=total_deadline)

    return primary

File: internal/providers/factory.go
package providers

import (
	"strings"
	"time"

	"github.com/rahul-k-r/auto-permissions-mode/internal/config"
)

// GetProvider wires the provider hierarchy based on application configuration.
func GetProvider(cfg config.Config) Provider {
	provName := strings.ToLower(strings.TrimSpace(cfg.Provider))
	endpoint := cfg.Endpoint
	if endpoint == "" {
		endpoint = "http://127.0.0.1:9931/v1/chat/completions"
	}
	model := cfg.Model
	timeout := time.Duration(cfg.TimeoutSeconds * float64(time.Second))
	if timeout <= 0 {
		timeout = 3500 * time.Millisecond
	}
	temp := cfg.Temperature
	numCtx := cfg.NumCtx
	if numCtx <= 0 {
		numCtx = 4096
	}
	maxTokens := cfg.MaxTokens
	if maxTokens <= 0 {
		maxTokens = 512
	}

	var primary Provider

	switch provName {
	case "gemini":
		apiKey := ResolveAPIKey("GEMINI_API_KEY", "gemini_api_key", cfg.GeminiAPIKey, cfg.APIKey)
		primary = NewGeminiProvider(apiKey, model, temp, timeout, maxTokens)
	case "anthropic":
		apiKey := ResolveAPIKey("ANTHROPIC_API_KEY", "anthropic_api_key", cfg.AnthropicAPIKey, cfg.APIKey)
		primary = NewAnthropicProvider(apiKey, model, temp, timeout, maxTokens)
	case "ollama":
		primary = NewOllamaProvider(endpoint, model, numCtx, temp, timeout, maxTokens)
	default:
		// llamacpp, openai, or generic openai-compatible
		apiKey := ResolveAPIKey("OPENAI_API_KEY", "openai_api_key", cfg.OpenAIAPIKey, cfg.APIKey)
		primary = NewOpenAICompatibleProvider(endpoint, model, apiKey, temp, timeout, maxTokens)
	}

	if cfg.FallbackToCloud {
		cloudProv := strings.ToLower(strings.TrimSpace(cfg.CloudProvider))
		cloudTimeout := time.Duration(cfg.CloudTimeoutSeconds * float64(time.Second))
		if cloudTimeout <= 0 {
			cloudTimeout = 12 * time.Second
		}
		cloudModel := cfg.CloudModel

		var secondary Provider
		switch cloudProv {
		case "anthropic":
			apiKey := ResolveAPIKey("ANTHROPIC_API_KEY", "anthropic_api_key", cfg.AnthropicAPIKey, cfg.APIKey)
			secondary = NewAnthropicProvider(apiKey, cloudModel, temp, cloudTimeout, maxTokens)
		case "openai", "openrouter", "groq":
			cloudEP := "https://api.openai.com/v1/chat/completions"
			switch cloudProv {
			case "openrouter":
				cloudEP = "https://openrouter.ai/api/v1/chat/completions"
			case "groq":
				cloudEP = "https://api.groq.com/openai/v1/chat/completions"
			}
			apiKey := ResolveAPIKey("OPENAI_API_KEY", "openai_api_key", cfg.OpenAIAPIKey, cfg.APIKey)
			if cloudProv == "openrouter" {
				if k := ResolveAPIKey("OPENROUTER_API_KEY", "openrouter_api_key", cfg.OpenRouterAPIKey, cfg.APIKey); k != "" {
					apiKey = k
				}
			}
			// Only OpenAI itself defaults to gpt-4o-mini; Groq/OpenRouter get "auto" so
			// the provider resolves an available model via the endpoint's /models list
			// instead of requesting a model name ("gpt-4o-mini") that endpoint doesn't serve.
			defModel := "auto"
			if cloudProv == "openai" {
				defModel = "gpt-4o-mini"
			}
			if cloudModel != "" {
				defModel = cloudModel
			}
			secondary = NewOpenAICompatibleProvider(cloudEP, defModel, apiKey, temp, cloudTimeout, maxTokens)
		default:
			// default gemini
			apiKey := ResolveAPIKey("GEMINI_API_KEY", "gemini_api_key", cfg.GeminiAPIKey, cfg.APIKey)
			secondary = NewGeminiProvider(apiKey, cloudModel, temp, cloudTimeout, maxTokens)
		}

		totalDeadline := time.Duration(cfg.TotalDeadlineSeconds * float64(time.Second))
		if totalDeadline <= 0 {
			totalDeadline = 18 * time.Second
		}
		return NewTieredProvider(primary, secondary, totalDeadline)
	}

	return primary
}
